src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
-        return introspectionData->second;
-
-    const ICPUBuffer* spv = params.cpuShader->getContent();
-    spirv_cross::Compiler comp(reinterpret_cast<const uint32_t*>(spv->getPointer()), spv->getSize()/4u);
-    auto introspection = doIntrospection(comp,params.entryPoint,params.cpuShader->getStage());
-    
-    if (insertToCache)
-        m_introspectionCache[params] = introspection;
-
-    return introspection;
+    switch (basetype)
+    {
+    case spirv_cross::SPIRType::Int:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::I32;
+    case spirv_cross::SPIRType::UInt:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::U32;
+    case spirv_cross::SPIRType::Float:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::F32;
+    case spirv_cross::SPIRType::Int64:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::I64;
+    case spirv_cross::SPIRType::UInt64:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::U64;
+    case spirv_cross::SPIRType::Double:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::F64;
+    default:
+        return CSPIRVIntrospector::CStageIntrospectionData::VAR_TYPE::UNKNOWN_OR_STRUCT;
+    }
 }
 
-bool CSPIRVIntrospector::introspectAllShaders(core::smart_refctd_ptr<const CIntrospectionData>* outIntrospections, const std::span<const ICPUShader::SSpecInfo> infos)
+// returns true if successfully added all the info to self, false if incompatible with what's already in our pipeline or incomplete (e.g. missing spec constants)
+NBL_API2 bool CSPIRVIntrospector::CPipelineIntrospectionData::merge(const CSPIRVIntrospector::CStageIntrospectionData* stageData, const ICPUShader::SSpecInfoBase::spec_constant_map_t* specConstants)
 {
-    auto it = outIntrospections;
-    for (auto info : infos)
+    // validate if descriptors are compatible
+    DescriptorSetBindings descriptorsToMerge[ICPUPipelineLayout::DESCRIPTOR_SET_COUNT];
+    for (uint32_t i = 0u; i < ICPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
     {
-        SIntrospectionParams params = {info.entryPoint,core::smart_refctd_ptr<const ICPUShader>(info.shader)};
-        *it = introspect(params);
-        if (!*it++)
-            return false;
-    }
-    return true;
-}
+        std::ostringstream debug;
+        debug << "ds ID: " << i << std::endl;
 
+        // TODO: test it with glsl
+        auto multiplyArrayDimensions = [&specConstants](uint32_t val, const SArrayInfo& arrInfo) -> uint32_t
+            {
+                if (arrInfo.value == 0)
+                    return val;
 
-template<E_SHADER_RESOURCE_TYPE restype>
-static std::pair<bool, IImageView<ICPUImage>::E_TYPE> imageInfoFromResource(const SShaderResource<restype>& _res)
-{
-    return {_res.shadow, _res.viewType};
-}
+                if (arrInfo.isSpecConstant)
+                {
+                    if (!specConstants)
+                        return 0u;
```

Comment:
you need to document that this lambda returning 0 indicates failure

Reply:
we are getting rid of `stride` anyway

Reply:
ah yes! need a new way to indicate failure I guess?

Reply:
done